### PR TITLE
feat: Hierarchical Partitioner to support multi-backends

### PR DESCRIPTION
**Pull request by zewenli98**
# Description

Hierarchical Partitioner to support multi-backends.

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: docs/hierarchical_partitioner.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
+# Hierarchical Partitioner for TensorRT
```
Comment (narendasan) [inline docs/hierarchical_partitioner.md:1]: Lets add this to the contributor docs for now

File: examples/hierarchical_partitioner_example.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                torch.ops.aten.conv2d.default,
+                torch.ops.aten.convolution.default,
+            },
+            "tensorrt": set(DYNAMO_ATEN_CONVERTERS.keys()),
+        },
+        torch_executed_ops=[
+            torch.ops.aten._native_batch_norm_legit_no_training.default
+        ],
+        require_full_compilation=False,
+        skip_fusion=False,
```
Comment (peri044) [inline examples/hierarchical_partitioner_example.py]: skip_fusion=False slows down the partitioning a lot. Can you check if it's really needed ?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 3]
```diff
@@ -913,26 +945,61 @@ def contains_metadata(gm: torch.fx.GraphModule) -> bool:
...
+                            torch.randn(input.shape)
+                            .to(dtype.to(input.dtype, t=torch.dtype))
+                            .cuda()
+                        )
+                        sub_inputs.append(sub_input)
+
+                    compiled_func = torch._inductor.compile(
+                        submodule,
+                        sub_inputs,
+                    )
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_compiler.py]: `torch.dynamo.mark_dynamic` API  is used to set dynamic shapes for torch.compile workflow. Reference: https://docs.pytorch.org/TensorRT/user_guide/dynamic_shapes.html. So you can use the construct_submodule_inputs() API to give you dynamic inputs (if that's the case) and set them to the inductor segment accordingly.

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 4]
```diff
@@ -913,26 +945,61 @@ def contains_metadata(gm: torch.fx.GraphModule) -> bool:
...
+                elif "_run_on_acc_tensorrt" in name:
+                    compiled_submodule = convert_module(
+                        submodule,
+                        submodule_inputs,
+                        settings=settings,
+                        name=name,
+                        engine_cache=engine_cache,
+                    )
+                else:
+                    raise ValueError(f"Unknown backend for submodule: {name}")
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_compiler.py]: should there be a `_run_on_gpu` segment here ? How would the torch.ops.aten._native_batch_norm_legit_no_training.default fallback to native pytorch in your example from slack ?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 5]
```diff
@@ -1276,3 +1343,12 @@ def load_cross_compiled_exported_program(file_path: str = "") -> Any:
...
     return replace_execute_engine_no_op_node(exp_program)
+
+
+class FunctionWrapper(torch.nn.Module):
+    def __init__(self, func):
+        super().__init__()
+        self.func = func
+
+    def forward(self, *args, **kwargs):
+        return self.func(*args, **kwargs)
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_compiler.py]: Consider naming this to `InductorModule` and moving to utils

File: py/torch_tensorrt/dynamo/partitioning/_hierarchical_partitioner.py [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+    partitioner = HierarchicalAdjacencyPartitioner(
+        gm,
+        supported_ops,
+        backend_support_map=backend_support_map,
+        backend_priority=backend_priority,
+        min_block_size=min_block_size,
+        require_full_compilation=require_full_compilation,
+        skip_fusion=skip_fusion,
+    )
+
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/partitioning/_hierarchical_partitioner.py:580]: Do you think copying all this code is necessary ? or inheriting only some components would be sufficient ?

**Review by peri044 — COMMENTED** [review]
Considering how this feature will be made available, the min_block_size and torch_executed_ops need to be re-thought or deprecated as the min_block_size can apply to all backends and `torch_executed_ops` will be replaced by `backend_support_map`

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 7]
```diff
@@ -913,26 +945,61 @@ def contains_metadata(gm: torch.fx.GraphModule) -> bool:
...
+                        sub_inputs.append(sub_input)
+
+                    compiled_func = torch._inductor.compile(
+                        submodule,
+                        sub_inputs,
+                    )
+                    # Wrap the compiled function to be a torch.nn.Module
+                    compiled_submodule = FunctionWrapper(compiled_func)
+
+                elif "_run_on_acc_tensorrt" in name:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: Is there some sort of design where the capability and conversion parts can be grouped and registered? We can add this concept to the RFC for later

Comment (narendasan) [reply py/torch_tensorrt/dynamo/_compiler.py]: That way we dont have a long conditional case set we just look up the appropriate conversion function on a standardized API

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 8]
```diff
@@ -913,26 +945,61 @@ def contains_metadata(gm: torch.fx.GraphModule) -> bool:
...
     parse_graph_io(gm, dryrun_tracker)
 
     # Replace all FX Modules with TRT Modules
-    for name, trt_module in trt_modules.items():
-        setattr(partitioned_module, name, trt_module)
+    for name, compiled_module in compiled_modules.items():
+        setattr(partitioned_module, name, compiled_module)
         if settings.lazy_engine_init and not settings.enable_cross_compile_for_windows:
-            getattr(partitioned_module, name).setup_engine()
+            if use_hierarchical_partitioner:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: Similar here, we should standardize post processing as well

**Review by narendasan — COMMENTED** [review]
What would be the merge-able version of this PR since we are going to expose the new partitioner yet?

File: examples/hierarchical_partitioner_example.py (hunk 2 above)
Comment (zewenli98) [reply examples/hierarchical_partitioner_example.py]: > the min_block_size and torch_executed_ops need to be re-thought or deprecated as the min_block_size can apply to all backends 

My understanding is that, if the num of ops of a GM is less than `min_block_size`, no matter it's TRT or other backend, the GM would not be compiled. Do you mean it shouldn't apply to other backends? 

> `torch_executed_ops` will be replaced by `backend_support_map`

My understanding is that `torch execute` is not considered as a backend because it doesn't need any compilation, it just runs ops in eager mode. So, if an op was in `torch_executed_ops`, it would ignore `backend_support_map` and run in torch eager anyway.

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 4 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/_compiler.py]: Nope, as I mentioned above, `_run_on_gpu` is not considered as a backend. We just keep the module as is. Only `_run_on_acc_backend` modules need to be replaced with the compiled module. This aligns with our existing implementation.

File: examples/hierarchical_partitioner_example.py (hunk 2 above)
Comment (zewenli98) [reply examples/hierarchical_partitioner_example.py]: > skip_fusion=False slows down the partitioning a lot. Can you check if it's really needed ?

Since adjacency partitioner uses this flag, I just keep it here. yeah I can definitely switch it to True in the example.

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 7 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/_compiler.py]: yeah I agree that the conversion part of different backends should be grouped, but for now I don't have too much info about other backends (like how to convert an op to that backend). We can definitely do this when we are ready to support other backends.

**Comment by zewenli98** [comment]
> What would be the merge-able version of this PR since we are going to expose the new partitioner yet?

@narendasan As you said that we don't want to expose APIs of this feature to users for now, the current code doesn't change any behavior. It can be merged after I resolve the issues above.

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 3 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/_compiler.py]: `submodule_inputs` is already the return of `construct_submodule_inputs()`. What I did here is to convert torch-trt Input to torch Tensor. I changed to:
```
sub_input = (
    input.torch_tensor
    .to(dtype.to(input.dtype, t=torch.dtype))
    .cuda()
)
```

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 9]
```diff
@@ -788,20 +798,47 @@ def contains_metadata(gm: torch.fx.GraphModule) -> bool:
             "Some nodes do not have metadata (shape and dtype information). This could lead to problems sometimes if the graph has PyTorch and TensorRT segments."
         )
 
+    ############ TODO: testing only ############
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: Should get removed right?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 10]
```diff
@@ -788,20 +798,47 @@ def contains_metadata(gm: torch.fx.GraphModule) -> bool:
...
-            logger.info("Partitioning the graph via the fast partitioner")
-            partitioned_module, supported_ops = partitioning.fast_partition(
-                gm,
-                verbose=settings.debug,
-                min_block_size=settings.min_block_size,
-                torch_executed_ops=settings.torch_executed_ops,
-                require_full_compilation=settings.require_full_compilation,
-                skip_fusion=(num_supported_ops == total_ops),
-            )
+            if use_hierarchical_partitioner:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py]: Should this be removed as well or should we just tie it to a set of hidden settings?

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 7 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/_compiler.py]: Lets remove the _hierarchical partitioner from _compile.py then and put together an RFC

**Review by narendasan — COMMENTED** [review]
Looks mostly good to me, just remove the stuff in _compile and make sure the example comprehensively demonstrates how to use the feature

**Comment by zewenli98** [comment]
> Looks mostly good to me, just remove the stuff in _compile and make sure the example comprehensively demonstrates how to use the feature

Should we expose APIs to users? If not, how can they use the feature?

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 9 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/_compiler.py]: If these variables are removed here, how to provide backend_priority and backend_support_map if we don't expose APIs?

File: examples/hierarchical_partitioner_example.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    partitioned_model, op_support = hierarchical_adjacency_partition(
+        gm,
+        verbose=True,
+        min_block_size=1,
+        backend_priority=["inductor", "tensorrt"],
+        backend_support_map={
+            "inductor": {
+                "torch.ops.aten.convolution.default",
+            },
+            "tensorrt": CONVERTERS.keys(),
```
Comment (peri044) [inline examples/hierarchical_partitioner_example.py]: is this needed ? this will also have the convolution op right ?

Comment (zewenli98) [reply examples/hierarchical_partitioner_example.py]: This is a demo just for showing the functionality. Since both inductor and tensorrt support conv and `backend_priority=["inductor", "tensorrt"]`, conv will run on inductor. The module structure:
```
Partitioned Model Structure:
 GraphModule(
  (_run_on_acc_inductor_0): GraphModule(
    (conv1): Module()
  )
  (_run_on_gpu_1): GraphModule(
    (bn1): Module()
  )
  (_run_on_acc_tensorrt_2): GraphModule()
  (_run_on_acc_inductor_3): GraphModule(
    (conv2): Module()
  )
  (_run_on_gpu_4): GraphModule(
    (bn2): Module()
  )
  (_run_on_acc_tensorrt_5): GraphModule()
)
```

File: py/torch_tensorrt/dynamo/partitioning/_hierarchical_partitioner.py (hunk 6 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/partitioning/_hierarchical_partitioner.py:580]: I revisited the code where it has only two minor functions are the same and one of them calls a customized function. Considering the maintainability and suggestions from Cursor, I think it would be better to separate them.

**Review by narendasan — APPROVED** [review]
LGTM
